src/Magnum/Platform/EmscriptenApplication.cpp

```diff
@@ -0,0 +1,707 @@
+                app.textInputEvent(e);
+                return e.isAccepted();
+            }
+            KeyEvent e{*event};
+            app.keyPressEvent(e);
+            return e.isAccepted();
+        }));
+
+    emscripten_set_keyup_callback(keyboardListeningElement, this, false,
+        ([](int, const EmscriptenKeyboardEvent* event, void* userData) -> Int {
+            KeyEvent e{*event};
+            static_cast<EmscriptenApplication*>(userData)->keyReleaseEvent(e);
+            return e.isAccepted();
+        }));
+
+    #ifdef EMSCRIPTEN_EVENT_TARGET_DOCUMENT
+    if(keyboardListeningElement != EMSCRIPTEN_EVENT_TARGET_DOCUMENT &&
+       keyboardListeningElement != EMSCRIPTEN_EVENT_TARGET_WINDOW)
+    #endif
+    {
+        std::free(keyboardListeningElement);
+    }
+}
+
+void EmscriptenApplication::startTextInput() {
+    _flags |= Flag::TextInputActive;
+}
+
+void EmscriptenApplication::stopTextInput() {
+    _flags &= ~Flag::TextInputActive;
+}
+
+void EmscriptenApplication::setTextInputRect(const Range2Di&) {
+    /** @todo place a hidden input field at given rect */
+}
+
+void EmscriptenApplication::viewportEvent(ViewportEvent&) {}
+void EmscriptenApplication::keyPressEvent(KeyEvent&) {}
+void EmscriptenApplication::keyReleaseEvent(KeyEvent&) {}
+void EmscriptenApplication::mousePressEvent(MouseEvent&) {}
+void EmscriptenApplication::mouseReleaseEvent(MouseEvent&) {}
+void EmscriptenApplication::mouseMoveEvent(MouseMoveEvent&) {}
+void EmscriptenApplication::mouseScrollEvent(MouseScrollEvent&) {}
+void EmscriptenApplication::textInputEvent(TextInputEvent&) {}
+
+#ifdef MAGNUM_TARGET_GL
+EmscriptenApplication::GLConfiguration::GLConfiguration():
+    _colorBufferSize{8, 8, 8, 0}, _depthBufferSize{24}, _stencilBufferSize{0} {}
+#endif
+
+int EmscriptenApplication::exec() {
+    redraw();
+    return 0;
+}
+
+void EmscriptenApplication::redraw() {
+    _flags |= Flag::Redraw;
+
+    /* Loop already running, no need to start */
+    if(!!(_flags & Flag::LoopActive)) return;
+
+    // TODO: @mosra There is a race condition potential here,
+    // if someone calls redraw() in async callback/event handler
+    // there could be chance that the above check fails again
+    // before the below is set...
+
+    /* Start requestAnimationFrame loop */
+    _flags |= Flag::LoopActive;
+
+    /* Callback type needs to be explicit, otherwise emscripten will
+       not cast to a function pointer properly.
+       Return false to cancel animation frame. */
+    int (*callback)(void*) = [](void* userData) -> int {
+        auto& app = *static_cast<EmscriptenApplication*>(userData);
+
+        if((app._flags & Flag::Redraw) && !(app._flags & (Flag::ExitRequested))) {
+            app._flags &= ~Flag::Redraw;
+            app.drawEvent();
+        }
+
+        /* If redraw is requested, we will not cancel the already requested
+           animation frame */
+        if(!!(app._flags & (Flag::Redraw)) && !(app._flags & (Flag::ExitRequested))) {
```

Comment:
```suggestion
        if(!!(app._flags & Flag::Redraw) && !(app._flags & Flag::ExitRequested)) {
```

(too) many braces, ((same above)) :trollface:

Reply:
Woups, that came from experimentation with trying to merge the two conditions into one &